CMakeLists.txt

```diff
@@ -0,0 +1,356 @@
+check_function_exists(select HAVE_SELECT)
+check_function_exists(socket HAVE_SOCKET)
+check_function_exists(strcasecmp HAVE_STRCASECMP)
+check_function_exists(pthread_mutex_lock HAVE_PTHREAD_MUTEX_LOCK)
+check_function_exists(getaddrinfo HAVE_GETADDRINFO)
+check_function_exists(strnlen HAVE_STRNLEN)
+check_function_exists(strrchr HAVE_STRRCHR)
+
+# check for symbols
+if(WIN32)
+  set(HAVE_STRUCT_CMSGHDR 1)
+else()
+  check_symbol_exists(cmsghdr sys/socket.h HAVE_STRUCT_CMSGHDR)
+endif()
+
+if(ENABLE_DTLS)
+  if(WITH_GNUTLS OR WITH_TINYDTLS)
+    set(WITH_OPENSSL OFF)
+  endif()
+
+  if(WITH_GNUTLS AND WITH_TINYDTLS)
+    message(FATAL_ERROR "cannot build with gnutls and tinydtls")
+  endif()
+
+  if(WITH_OPENSSL)
+    set(WITH_OPENSSL "${WITH_OPENSSL}")
+    message(STATUS "compiling with openssl support")
+    find_package(OpenSSL REQUIRED)
+    set(HAVE_OPENSSL 1)
+  endif()
+
+  if(WITH_GNUTLS)
+    set(WITH_GNUTLS "${WITH_GNUTLS}")
+    message(STATUS "compiling with gnutls support")
+    # required libgnutls (e.g. debian libgnutls28-dev)
+    find_package(GnuTLS REQUIRED)
+    set(HAVE_LIBGNUTLS 1)
+  endif()
+
+  if(WITH_TINYDTLS)
+    set(WITH_TINYDTLS "${WITH_TINYDTLS}")
+    message(STATUS "compiling with tinydtls support")
+    set(HAVE_LIBTINYDTLS 1)
+  endif()
+endif()
+
+message(STATUS "ENABLE_DTLS: ${ENABLE_DTLS}")
+message(STATUS "WITH_TINYDTLS: ${WITH_TINYDTLS}")
+message(STATUS "WITH_GNUTLS: ${WITH_GNUTLS}")
+message(STATUS "WITH_OPENSSL: ${WITH_OPENSSL}")
+message(STATUS "HAVE_LIBTINYDTLS: ${HAVE_LIBTINYDTLS}")
+message(STATUS "HAVE_LIBGNUTLS: ${HAVE_LIBGNUTLS}")
+message(STATUS "HAVE_OPENSSL: ${HAVE_OPENSSL}")
+
+if(${WITH_EPOLL} AND ${HAVE_EPOLL_H} AND ${HAVE_TIMERFD_H})
+  set(COAP_EPOLL_SUPPORT "1")
+  message(STATUS "compiling with epoll support")
+endif()
+
+if(ENABLE_SMALL_STACK)
+  set(ENABLE_SMALL_STACK "${ENABLE_SMALL_STACK}")
+  message(STATUS "compiling with small stack support")
+endif()
+
+set(PACKAGE_URL "https://libcoap.net/")
+set(PACKAGE_NAME "${PROJECT_NAME}")
+set(PACKAGE_TARNAME "${PROJECT_NAME}")
+set(PACKAGE_STRING "${PROJECT_NAME} ${PROJECT_VERSION}")
+set(PACKAGE_VERSION "${PROJECT_VERSION}")
+set(LIBCOAP_PACKAGE_VERSION "${PROJECT_VERSION}")
+set(LIBCOAP_PACKAGE_URL "${PACKAGE_URL}")
+set(LIBCOAP_PACKAGE_NAME "${PROJECT_NAME}")
+set(LIBCOAP_PACKAGE_STRING "TODO")
+set(LIBCOAP_PACKAGE_BUGREPORT "libcoap-developers@lists.sourceforge.net")
+set(PACKAGE_BUGREPORT "libcoap-developers@lists.sourceforge.net")
+set(top_srcdir "${CMAKE_CURRENT_LIST_DIR}")
+set(top_builddir "${CMAKE_CURRENT_BINARY_DIR}")
+
+# creates config header file in build directory
+configure_file(${CMAKE_CURRENT_LIST_DIR}/include/coap2/coap.h.in
+               ${CMAKE_CURRENT_BINARY_DIR}/include/coap2/coap.h)
+
+configure_file(${CMAKE_CURRENT_LIST_DIR}/include/coap2/coap_config.h.in
+               ${CMAKE_CURRENT_BINARY_DIR}/include/coap2/coap_config.h)
+
+#
+# library
+#
+
+add_library(
+  ${PROJECT_NAME}
```

Comment:
This only builds static libraries - we also need dynamically loadable libraries as a configurable option or document somewhere that BUILD_SHARED_LIBS is ON is required.

Reply:
if you need shared libraries: 
```bash
rm -rf _build
mkdir _build
cd _build
cmake .. -DBUILD_SHARED_LIBS=ON
cmake --build . -j
ls -al _build/ | grep .so
```